### PR TITLE
[Model] Add Cosmos Predict 2.5 support.

### PR DESCRIPTION
**Pull request by oscardev256**
## Purpose
Add Cosmos-Predict2.5 model (text2world, image2world, and video2world) support to address issue #1747

This PR covers single-GPU offline inference. Advanced features and Cosmos-Transfer2.5 are left for follow-up PRs.

The only change beyond the Cosmos-Predict2.5 model is passing `revision` through the model loader (`diffusion/data.py`, `engine/async_omni_engine.py`, `entrypoints/omni_base.py`, `entrypoints/utils.py`), since the weights live on a non-default HF branch (`diffusers/base/post-trained`).

 | Size | Pipeline class                  | Supported tasks                          |
 |------|---------------------------------|------------------------------------------|
 | 2B   | `Cosmos2_5_PredictBasePipeline` | Text2World, Image2World, Video2World     |
 | 14B  | `Cosmos2_5_PredictBasePipeline` | Text2World, Image2World, Video2World     |

## Test Plan
End-to-end manual runs against the upstream diffusers `Cosmos2_5_PredictBasePipeline` for visual parity. Same prompt / seed / resolution per pair.

  **Text2World — 2B**

```bash
python examples/offline_inference/cosmos/cosmos_predict2_5.py \
--mode text2world \
--model nvidia/Cosmos-Predict2.5-2B \
--revision diffusers/base/post-trained \
--prompt "As the red light shifts to green, the red bus at the intersection begins to move forward, its headlights cutting through the falling snow. The snowy tire tracks deepen as the vehicle inches ahead, casting fresh lines onto the slushy road. Around it, streetlights glow warmer, illuminating the drifting flakes and wet reflections on the asphalt. Other cars behind start to edge forward, their beams joining the scene. The stillness of the urban street transitions into motion as the quiet snowfall is punctuated by the slow advance of traffic through the frosty city corridor." \
--num-frames 93 --seed 1 \
--output cosmos_predict_2_5_2B_t2v_out.mp4
```

  Text2World — 14B (CPU offload)

```bash
python examples/offline_inference/cosmos/cosmos_predict2_5.py \
--mode text2world \
--model nvidia/Cosmos-Predict2.5-14B \
--revision diffusers/base/post-trained \
--prompt "As the red light shifts to green, the red bus at the intersection begins to move forward, its headlights cutting through the falling snow. The snowy tire tracks deepen as the vehicle inches ahead, casting fresh lines onto the slushy road. Around it, streetlights glow warmer, illuminating the drifting flakes and wet reflections on the asphalt. Other cars behind start to edge forward, their beams joining the scene. The stillness of the urban street transitions into motion as the quiet snowfall is punctuated by the slow advance of traffic through the frosty city corridor." \
--num-frames 93 --seed 1 \
--enable-cpu-offload \
--height 480 --width 832 \
--output cosmos_predict_2_5_14B_t2v_out.mp4
```

  Image2World — 2B

```bash
python examples/offline_inference/cosmos/cosmos_predict2_5.py \
--mode image2world \
--model nvidia/Cosmos-Predict2.5-2B \
--revision diffusers/base/post-trained \
--prompt "A nighttime city bus terminal gradually shifts from stillness to subtle movement. At first, multiple double-decker buses are parked under the glow of overhead lights, with a central bus labeled '87D' facing forward and stationary. As the video progresses, the bus in the middle moves ahead slowly, its headlights brightening the surrounding area and casting reflections onto adjacent vehicles. The motion creates space in the lineup, signaling activity within the otherwise quiet station. It then comes to a smooth stop, resuming its position in line. Overhead signage in Chinese characters remains illuminated, enhancing the vibrant, urban night scene." \
--image ./bus_terminal.jpg \
--num-frames 93 --seed 1 \
--output cosmos_predict_2_5_2B_i2v_out.mp4
```

  Video2World — 2B

```sh
python examples/offline_inference/cosmos/cosmos_predict2_5.py \
--mode video2world \
--model nvidia/Cosmos-Predict2.5-2B \
--revision diffusers/base/post-trained \
--prompt "A robotic arm, primarily white with black joints and cables, is shown in a clean, modern indoor setting with a white tabletop. The arm, equipped with a gripper holding a small, light green pitcher, is positioned above a clear glass containing a reddish-brown liquid and a spoon. The robotic arm is in the process of pouring a transparent liquid into the glass. To the left of the pitcher, there is an opened jar with a similar reddish-brown substance visible through its transparent body. In the background, a vase with white flowers and a brown couch are partially visible, adding to the contemporary ambiance. The lighting is bright, casting soft shadows on the table. The robotic arm's movements are smooth and controlled, demonstrating precision in its task. As the video progresses, the robotic arm completes the pour, leaving the glass half-filled with the reddish-brown liquid. The jar remains untouched throughout the sequence, and the spoon inside the glass remains stationary. The other robotic arm on the right side also stays stationary throughout the video. The final frame captures the robotic arm with the pitcher finishing the pour, with the glass now filled to a higher level, while the pitcher is slightly tilted but still held securely by the gripper."
--video ./robot_pouring_1.mp4 \
--num-frames 93 --seed 1 \
--output cosmos_predict_2_5_2B_v2w_out.mp4
```

## Test Result

<strong>Text2World — 2B</strong>

<details>
<summary>Details</summary>

<table>
<tr>
<th>vLLM-Omni</th>
<th>Diffusers Reference</th>
</tr>

<tr>
<td>
<video src="https://github.com/user-attachments/assets/437d56c6-60cb-4330-b663-946c19759778"
controls width="420"></video>
</td>

<td>
<video src="https://github.com/user-attachments/assets/5cd59a1a-8854-4cd3-aeff-fdf713ddbb71"
controls width="420"></video>
</td>
</tr>
</table>

</details>

<strong>Text2World — 14B</strong>

<details>
<summary>Details</summary>

<table>
<tr>
<th>vLLM-Omni</th>
<th>Diffusers Reference</th>
</tr>

<tr>
<td>
<video src="https://github.com/user-attachments/assets/d1ee7c15-d8eb-4f86-aa83-7c042332ea4b"
controls width="420"></video>
</td>

<td>
<video src="https://github.com/user-attachments/assets/153f3275-40f2-407c-a4d1-628c1ae39f41"
controls width="420"></video>
</td>
</tr>
</table>

</details>

<strong>Image2World — 2B</strong>

<details>
<summary>Details</summary>

<table>
<tr>
<th>vLLM-Omni</th>
<th>Diffusers Reference</th>
</tr>

<tr>
<td>
<video src="https://github.com/user-attachments/assets/ff9918ae-1c6b-4095-bebe-18b8f629a4e2"
controls width="420"></video>
</td>

<td>
<video src="https://github.com/user-attachments/assets/9e8e265f-ffb3-47a5-a884-1bb53bfe555a"
controls width="420"></video>
</td>
</tr>
</table>

</details>

<strong>Video2World — 2B</strong>

<details>
<summary>Details</summary>

<table>
<tr>
<th>vLLM-Omni</th>
<th>Diffusers Reference</th>
</tr>

<tr>
<td>
<video src="https://github.com/user-attachments/assets/3b590c3d-2399-4b16-ad1a-5406b7d7d4a9"
controls width="420"></video>
</td>

<td>
<video src="https://github.com/user-attachments/assets/9901911e-8151-473d-bba0-ae2dc36af9e2"
controls width="420"></video>
</td>
</tr>
</table>

</details>

## Memory & Latency

Measured on a single NVIDIA A100-SXM4-80GB at default settings (704×1280, 93 frames, 36 steps):

| Model | Mode | Flags | Peak VRAM (BF16) | Wall time |
|---|---|---|---|---|
| `Cosmos-Predict2.5-2B` | Text2World | — | 36.6 GiB | ~13 min |
| `Cosmos-Predict2.5-2B` | Image2World | — | 35.5 GiB | ~13 min |
| `Cosmos-Predict2.5-2B` | Video2World | — | 36.5 GiB | ~13 min |
| `Cosmos-Predict2.5-14B` | Text2World | `--enable-cpu-offload` | 49.2 GiB | ~48 min |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [x] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/cosmos/cosmos_transformer.py [hunk 1]
```diff
@@ -0,0 +1,922 @@
...
+            for param_name, weight_name, shard_id in stacked_params_mapping:
+                if weight_name not in original_name:
+                    continue
+                lookup_name = original_name.replace(weight_name, param_name)
+                param = params_dict[lookup_name]
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
+                if ".ffn.net.0." in lookup_name:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/cosmos/cosmos_transformer.py]: This remaps `.ffn.net.0.` but the block attribute is `self.ff` (line 596), so `named_parameters()` produces `.ff.net.0.` keys. This remapping will never match and FFN weights will silently fail to load.

Looks copy-pasted from the Wan loader — either remove it (if checkpoint keys already use `ff.net.0.`) or rename the attribute to `ffn`.

File: vllm_omni/diffusion/models/cosmos/__init__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+__all__ = [
+    "CosmosTransformer3DModel",
+    "CosmosPredict25Pipeline",
+    "get_cosmos_predict25_post_process_func",
+    "retrieve_latents",
+    "load_transformer_config",
+    "create_transformer_from_config",
+    "CosmosTransformer3DModel",    
+]
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/cosmos/__init__.py:11]: `retrieve_latents`, `load_transformer_config`, `create_transformer_from_config` are in `__all__` but never imported, and `CosmosTransformer3DModel` is duplicated. This will break `from cosmos import *`.

File: vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+            )
+            with open(config_path) as f:
+                return json.load(f)
+        except Exception:
+            pass
+    return {}
+
+
+def create_transformer_from_config(config: dict) -> CosmosTransformer3DModel:
+    """Create WanTransformer3DModel from config dict."""
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py]: Docstring says "Create WanTransformer3DModel" and the kwargs (`text_dim`, `freq_dim`, `ffn_dim`, `cross_attn_norm`, `image_dim`, `added_kv_proj_dim`, etc.) are Wan-specific — `CosmosTransformer3DModel.__init__` doesn't accept them. This function is unused (you call the constructor directly at line 244). Just remove it.

File: vllm_omni/diffusion/models/cosmos/cosmos_transformer.py [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+    """
+
+    def __init__(self, in_features: int, out_features: int):
+        super().__init__()
+        self.linear_1 = nn.Linear(in_features, out_features, bias=False)
+        self.activation = nn.SiLU()
+        self.linear_2 = nn.Linear(out_features, 3 * out_features, bias=False)
+
+    def forward(self, timesteps: torch.Tensor) -> torch.Tensor:
+        emb = self.linear_1(timesteps.to(torch.bfloat16))
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/cosmos/cosmos_transformer.py]: Nit: hardcoded `torch.bfloat16` — should use the module's dtype.

```suggestion
        emb = self.linear_1(timesteps.to(self.linear_1.weight.dtype))
```

File: vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+                "[NVIDIA Open Model License Agreement](https://www.nvidia.com/en-us/agreements/enterprise-software/nvidia-open-model-license). "
+                f"Please ensure that you are compliant with the license agreement."
+            )
+
+        # Get parameters from request or arguments
+        if len(req.prompts) == 1:  # If req.prompt is empty, default to prompt & neg_prompt in param list
+            prompt = req.prompts[0] if isinstance(req.prompts[0], str) else req.prompts[0].get("prompt")
+            negative_prompt = None if isinstance(req.prompts[0], str) else req.prompts[0].get("negative_prompt")
+        if prompt is None and prompt_embeds is None:
+            raise ValueError("Prompt or prompt_embeds is required for Wan2.2 generation.")
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py]:
```suggestion
            raise ValueError("Prompt or prompt_embeds is required for Cosmos Predict 2.5 generation.")
```

**Review by lishunyang12 — CHANGES_REQUESTED** [review]
A few issues — the weight loading mapping looks broken and there's leftover Wan-specific code throughout.

File: vllm_omni/diffusion/models/cosmos/__init__.py (hunk 2 above)
Comment (oscardev256) [reply vllm_omni/diffusion/models/cosmos/__init__.py:11]: Fixed. __all__ now lists only the three symbols actually imported (`CosmosTransformer3DModel,` `CosmosPredict25Pipeline`, ` get_cosmos_predict25_post_process_func`) with duplicates removed.

File: vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py (hunk 3 above)
Comment (oscardev256) [reply vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py]: Fixed. Function removed.

File: vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py (hunk 5 above)
Comment (oscardev256) [reply vllm_omni/diffusion/models/cosmos/pipeline_cosmos_predict2_5.py]: Updated with proper model name.

File: vllm_omni/diffusion/models/cosmos/cosmos_transformer.py (hunk 4 above)
Comment (oscardev256) [reply vllm_omni/diffusion/models/cosmos/cosmos_transformer.py]: Removed the hardcoded dtype. `timesteps` is already cast to the hidden-states dtype upstream (CosmosEmbedding), so the embedding now follows the module dtype.

File: vllm_omni/diffusion/models/cosmos/cosmos_transformer.py (hunk 1 above)
Comment (oscardev256) [reply vllm_omni/diffusion/models/cosmos/cosmos_transformer.py]: Fixed. Removed this remapping.

**Comment by oscardev256** [comment]
Closing this PR due to Cosmos 2.5 being sunset to focus on Cosmos 3 per community discussion. Thanks all for the reviews.
